homeassistant/components/modem_callerid/sensor.py

```diff
@@ -11,50 +10,88 @@
-DEFAULT_DEVICE = "/dev/ttyACM0"
-
-STATE_RING = "ring"
-STATE_CALLERID = "callerid"
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-        vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
-    }
+PLATFORM_SCHEMA = cv.deprecated(
+    vol.All(
+        PLATFORM_SCHEMA.extend(
+            {
+                vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+                vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+            }
+        )
+    )
 )
 
+_LOGGER = logging.getLogger(__name__)
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up modem caller ID sensor platform."""
 
-    name = config.get(CONF_NAME)
-    port = config.get(CONF_DEVICE)
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Modem Caller ID sensor."""
+    try:
+        name = entry.data[CONF_NAME]
+    except KeyError:
+        name = DEFAULT_NAME
+    try:
+        device = entry.data[CONF_DEVICE]
+    except KeyError:
+        device = DEFAULT_DEVICE
+    api = hass.data[DOMAIN][entry.entry_id][DATA_KEY_API]
+    async_add_entities(
+        [
+            ModemCalleridSensor(
+                hass,
+                api,
+                name,
+                device,
+                entry.entry_id,
+            )
+        ]
+    )
+
+    async def on_hass_stop(self, event):
+        """HA is shutting down, close modem port."""
+        if self.api:
+            self.api.close()
+            self.api = None
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
+    )
 
-    modem = bm(port)
-    if modem.state == modem.STATE_FAILED:
-        _LOGGER.error("Unable to initialize modem")
-        return
+    platform = entity_platform.current_platform.get()
 
-    add_entities([ModemCalleridSensor(hass, name, port, modem)])
+    platform.async_register_entity_service(SERVICE_REJECT_CALL, {}, "reject_call")
 
 
 class ModemCalleridSensor(SensorEntity):
     """Implementation of USB modem caller ID sensor."""
 
-    def __init__(self, hass, name, port, modem):
+    def __init__(self, hass, api, name, device, server_unique_id):
         """Initialize the sensor."""
         self._attributes = {"cid_time": 0, "cid_number": "", "cid_name": ""}
-        self._name = name
-        self.port = port
-        self.modem = modem
+        self._device_class = None
+        self.device = device
+        self.api = api
         self._state = STATE_IDLE
-        modem.registercallback(self._incomingcallcallback)
-        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self._stop_modem)
+        self._name = name
+        self._server_unique_id = server_unique_id
+
+    async def async_added_to_hass(self):
+        """Call when the modem sensor is added to Home Assistant."""
+        self.api.registercallback(self._incomingcallcallback)
```

Comment:
```suggestion
        self.api.registercallback(self._async_incoming_call)
```

Please use snake case